wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,542 @@
+		}
+		if k.client.slot == 0 {
+			k.client.slot = opts.withSlot
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			k.client.pin = os.Getenv("PKCS11_PIN")
+		}
+		if k.client.pin == "" {
+			k.client.pin = opts.withPin
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			k.client.module = os.Getenv("PKCS11_MODULE")
+		}
+		if k.client.module == "" {
+			k.client.module = opts.withModule
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			k.client.label = os.Getenv("PKCS11_LABEL")
+		}
+		if k.client.label == "" {
+			k.client.label = opts.withLabel
+		}
+
+		if !opts.Options.WithDisallowEnvVars {
+			mechanismName := os.Getenv("PKCS11_MECHANISM")
+			if mechanismName != "" {
+				k.client.mechanism, err = MechanisFromString(mechanismName)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		if k.client.mechanism == 0 {
+			if opts.withMechanism != "" {
+				k.client.mechanism, err = MechanisFromString(opts.withMechanism)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		k.client.keyId = k.keyId
+
+		p := pkcs11.New(k.client.module)
+		err := p.Initialize()
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize PKCS11: %w", err)
+		}
+		defer p.Destroy()
+		defer p.Finalize()
+
+		session, err := p.OpenSession(k.client.slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open session: %w", err)
+		}
+		defer p.CloseSession(session)
+
+		err = p.Login(session, pkcs11.CKU_USER, k.client.pin)
+		if err != nil {
+			return nil, fmt.Errorf("failed to login: %w", err)
+		}
+		defer p.Logout(session)
+
+	}
+	// Store the current key id. If using a key alias, this will point to the actual
+	// unique key that that was used for this encrypt operation.
+	k.currentKeyId.Store(k.keyId)
+
+	// Map that holds non-sensitive configuration info
+	wrapConfig := new(wrapping.WrapperConfig)
+	wrapConfig.Metadata = make(map[string]string)
+	wrapConfig.Metadata["kms_key_id"] = k.keyId
+	wrapConfig.Metadata["slot"] = strconv.Itoa(int(k.client.slot))
+	if k.client.label != "" {
+		wrapConfig.Metadata["label"] = k.client.label
+	}
+	if k.client.mechanism != 0 {
+		wrapConfig.Metadata["mechanism"] = MechanisString(k.client.mechanism)
+	}
+
+	return wrapConfig, nil
+}
+
+// Type returns the type for this particular wrapper implementation
+func (k *Wrapper) Type(_ context.Context) (wrapping.WrapperType, error) {
+	return wrapping.WrapperTypePkcs11, nil
+}
+
+// KeyId returns the last known key id
+func (k *Wrapper) KeyId(_ context.Context) (string, error) {
+	return k.currentKeyId.Load().(string), nil
+}
+
+// Encrypt is used to encrypt the master key using the the PKCS11.
+// This returns the ciphertext, and/or any errors from this
+// call. This should be called after the KMS client has been instantiated.
+func (k *Wrapper) Encrypt(_ context.Context, plaintext []byte, opt ...wrapping.Option) (*wrapping.BlobInfo, error) {
+	if plaintext == nil {
```

Comment:
I guess this is the check used elsewhere, but I'd probably prefer:

```suggestion
	if len(plaintext) == 0 {
```